### PR TITLE
working new item button for nested (> 2nd level) inline collections

### PR DESCRIPTION
**Pull request by dypa**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #262, #1971 , #1229, #1999
| License       | MIT
| Doc PR        | no

### CODE REVIEW

**Comment by ehibes** [comment]
+1 We need this feature.

**Comment by seb0zz** [comment]
+1 would be so awesome

**Comment by huitiemesens** [comment]
:+1:

**Comment by AlexandreKilian** [comment]
+1, really need that currently

**Comment by taymour** [comment]
+1 Awesome feature

**Comment by snarky** [comment]
:+1:

**Comment by fredericlam** [comment]
+1, can't wait this feature !

**Comment by meniku** [comment]
please merge this!

**Comment by nio-dtp** [comment]
+1 to merge this awesome feature !

**Comment by krewetka** [comment]
:+1: ...it's been more than a year already and it's still not merged :-/ 

It's really important feature and many people voted for it. As @nicodmf wrote in December:

"The PR is ready for the developper, it would be gently to take a look on this good work : Tests passed, many people want this feature. Could you please review it ?"

@rande - is there any chance it will be reviewed and added to 2.3 version?

**Comment by greg0ire** [comment]
"This pull request contains merge conflicts that must be resolved"

It also contains a merge commit not from @dypa

**Comment by krewetka** [comment]
@greg0ire - ok, so what? What do you want to say?

It's commit from @rande - so shouldn't be a problem for him to merge it.

**Comment by greg0ire** [comment]
The P.R. should be rebased to resolve the merge conflict, and ideally, P.R. should not contain commits that already are in the tree (that's a bit weird isn't it ?)

**Comment by dypa** [comment]
>The P.R. should be rebased to resolve the merge conflicts.
>that's a bit weird isn't it ?
I rebased this pull request ~ 6 month ago and i have not received any response for 6 month. "Money in the morning, chairs in the evening" (c)

**Comment by greg0ire** [comment]
I understand your feeling (experienced that myself). You can't do much in that case, apart trying to get the P.R. cleaner and cleaner so that the day the maintainer has time to look at it, he doesn't delay it to another 6 months later because of something like that. Thank you for trying to get that merged, by the way.

**Comment by krewetka** [comment]
" P.R. should not contain commits that already are in the tree" - you are right :) 

Haven't realized that is was already merged somewhere else in a tree. 

Still I hope that at some point someone will merge this PR :)

**Comment by amine26tz** [comment]
@krewetka, you can see that the merge button is disabled, cause of conflicts ;). 
@dypa you should rebase this PR and resolve conflict

**Comment by greg0ire** [comment]
What is strange is that this merge commit merges something by pokap , so it really has nothing to do with this P.R.

I hope this P.R. would get merged too, and the less shady it will look, the more likely it is to get merged. Sorry for @dypa , you probably did not plan on having to maintain a P.R. Please note that if anyone wants to help @dypa do that, they can : 

1. clone his repo
2. add the main repo as a remote
3. do the rebase
4. publish the rebase branch on their fork
5. ask @dypa to replace his branch with theirs.

**Comment by dypa** [comment]
The problem is not in the Merge button, rebase or conflicts. If you want to help me then clone repo/use patch, check that it works not only for me (example https://github.com/dypa/sonata-admin-sandbox/tree/issue1971) and send feedback!

**Comment by greg0ire** [comment]
> The problem is not in the Merge button, rebase or conflicts.

not **only** 

If there are conflicts, this is not going to get merged at all (theoretically, it could, but it is a bit more involving than clicking a button). Check with the maintainers, but I doubt it is their policy to merge branch that can't be fast-forwarded.

**Comment by webdevilopers** [comment]
:+1:

**Comment by fre5h** [comment]
+1 for this feature

**Comment by ajouve** [comment]
:+1:

**Comment by sfavot** [comment]
+10000

**Comment by spyl94** [comment]
:+1:

**Comment by LogansUA** [comment]
:+1:

**Comment by stetodd** [comment]
:+1:

**Comment by sfavot** [comment]
@dypa what can we do to help?

**Comment by jorgelive** [comment]
@dypa:

I'm getting an error, replacing the original file with your git's file,

Error: Call to a member function getAssociationAdmin() on null 

if (count($elements) > 0) {
$elementId = array_shift($elements);
$fieldDescription = $admin->getFormFieldDescription($elementId);
$recursionAdmin = $fieldDescription->getAssociationAdmin(); ---->>> line with error
if ($fieldDescription && $recursionAdmin) {
I'm using:
symfomy 2.6-dev, admin-bundle dev-master

**Comment by greg0ire** [comment]
@jorgelive : if you want this to get merged one day, solve the merge conflicts. This is not a sufficient condition, but it is necessary.

**Comment by jorgelive** [comment]
@greg0ire : It's done, I hope semeday this can be implemented.

**Comment by greg0ire** [comment]
Great! Hopefully #2985 will be merged soon after 2.4 is released.

**Comment by muka** [comment]
:+1:

**Comment by Emirii** [comment]
Hello, I hope that this will be fixed soon. I am having a problem with the above hotfix however (AdminHelper.php + HelperController.php)

```
PropertyAccessor requires a graph of objects or arrays to operate on, but it found type "boolean" while trying to traverse path "phones[0]" at property "phones".
```

I have entities that look like this:
Client < Ads < Phones
Client < Ads < Addresses

Client can have multiple Ads, and Ad can have multiple Phones and Addresses. In my Sonata Admin I have them all inline. When I click "Add Ad" it appears fine, when i click "Add Phone" under that Ad it works fine, but when I click "Add Phone" again, or try to click "Add Addresses"  it gives me the above error. It actually gives me the above error no matter which "Add" button I pick, 1st or 2nd level.

Any tips appreciated, using the most recent dev pull.

File: Admin/AdminHelper.php [hunk 1]
```diff
@@ -32,25 +28,6 @@ public function __construct(Pool $pool)
...
 
     /**
-     * @throws \RuntimeException
-     *
-     * @param \Symfony\Component\Form\FormBuilder $formBuilder
-     * @param string                              $elementId
-     *
-     * @return \Symfony\Component\Form\FormBuilder
-     */
-    public function getChildFormBuilder(FormBuilder $formBuilder, $elementId)
```
Comment (Soullivaneuh) [inline Admin/AdminHelper.php]: Removing a public method is completely BC break.

BTW, why removing it?

If it's necessary, please make BC with deprecation system.

File: Admin/AdminHelper.php [hunk 2]
```diff
@@ -96,102 +68,57 @@ public function getAdmin($code)
...
      *
-     * @param object                                              $object
-     * @param \Sonata\AdminBundle\Admin\FieldDescriptionInterface $fieldDescription
+     * @param AdminInterface $admin
+     * @param string         $elementId
      *
      * @throws \RuntimeException
      */
-    public function addNewInstance($object, FieldDescriptionInterface $fieldDescription)
+    protected function addNewInstance(AdminInterface $admin, $elementId)
```
Comment (Soullivaneuh) [inline Admin/AdminHelper.php]: Changing method signature is BC break too.

File: Admin/AdminHelper.php [hunk 3]
```diff
@@ -96,102 +68,57 @@ public function getAdmin($code)
...
-     * Camelize a string
-     *
-     * @static
-     *
-     * @param string $property
-     *
-     * @return string
-     */
-    public function camelize($property)
+    protected function entityClassNameFinder(AdminInterface $admin, $elements)
```
Comment (Soullivaneuh) [inline Admin/AdminHelper.php]: BC break. Same reasons.

**Comment by greg0ire** [comment]
@Soullivaneuh : this should be closed in favor of #2985 , shouldn't it ?

**Comment by Soullivaneuh** [comment]
Indeed, thanks @greg0ire.

**Comment by kenanduman1988** [comment]
+1000

**Comment by greg0ire** [comment]
Locking as closed in favor of #2985 and people think it's ok to periodically come spam us with +10**x comments

**Pull request closed without merging**